systest/tests/distributed_post_verification_test.go

```diff
@@ -0,0 +1,298 @@
+package tests
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	grpc_logsettable "github.com/grpc-ecosystem/go-grpc-middleware/logging/settable"
+	grpczap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
+	"github.com/libp2p/go-libp2p/core/peer"
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"github.com/spacemeshos/post/shared"
+	"github.com/spacemeshos/post/verifying"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/activation"
+	"github.com/spacemeshos/go-spacemesh/api/grpcserver"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/config"
+	"github.com/spacemeshos/go-spacemesh/config/presets"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/p2p/handshake"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/localsql"
+	"github.com/spacemeshos/go-spacemesh/systest/cluster"
+	"github.com/spacemeshos/go-spacemesh/systest/testcontext"
+	"github.com/spacemeshos/go-spacemesh/timesync"
+	"github.com/spacemeshos/go-spacemesh/timesync/peersync"
+)
+
+var grpclog grpc_logsettable.SettableLoggerV2
+
+func init() {
+	grpclog = grpc_logsettable.ReplaceGrpcLoggerV2()
+}
+
+func TestPostMalfeasanceProof(t *testing.T) {
+	t.Parallel()
+	testDir := t.TempDir()
+
+	ctx := testcontext.New(t, testcontext.Labels("sanity"))
+	logger := ctx.Log.Desugar().WithOptions(zap.IncreaseLevel(zapcore.InfoLevel), zap.WithCaller(false))
+
+	// Prepare cluster
+	ctx.PoetSize = 1 // one poet guarantees everybody gets the same proof
+	ctx.ClusterSize = 3
+	cl := cluster.New(ctx, cluster.WithKeys(10))
+	require.NoError(t, cl.AddBootnodes(ctx, 1))
+	require.NoError(t, cl.AddBootstrappers(ctx))
+	require.NoError(t, cl.AddPoets(ctx))
+	require.NoError(t, cl.AddSmeshers(ctx, ctx.ClusterSize-cl.Total(), cluster.WithFlags(cluster.PostK3(1))))
+
+	// Prepare config
+	cfg, err := presets.Get("fastnet")
+	require.NoError(t, err)
+	cfg.Genesis = &config.GenesisConfig{
+		GenesisTime: cl.Genesis().Format(time.RFC3339),
+		ExtraData:   cl.GenesisExtraData(),
+	}
+	cfg.LayersPerEpoch = uint32(testcontext.LayersPerEpoch.Get(ctx.Parameters))
+	types.SetLayersPerEpoch(cfg.LayersPerEpoch)
+	cfg.LayerDuration = testcontext.LayerDuration.Get(ctx.Parameters)
+
+	cfg.DataDirParent = testDir
+	cfg.SMESHING.Opts.DataDir = filepath.Join(testDir, "post-data")
+	cfg.P2P.DataDir = filepath.Join(testDir, "post-data")
+	require.NoError(t, os.Mkdir(cfg.P2P.DataDir, os.ModePerm))
```

Comment:
Using the NodeID here instead of the p2p ID would simplify the test setup 🙂

Reply:
I don't understand

Reply:
After discussing in more detail what the source of randomness should be I'm ok with using p2p ID.